esp/esdllib/esdl_script.cpp

```diff
@@ -2310,6 +2340,318 @@ class CEsdlTransformOperationFunction : public CEsdlTransformOperationWithChildr
+    {
+        IEspContext* espCtx = scriptContext->queryEspContext();
+        CTxSummary* txSummary = (espCtx ? espCtx->queryTxSummary() : nullptr);
+        if (!txSummary)
+            return true;
+        StringBuffer value;
+        OptionalCriticalBlock block(crit);
+        unsigned start = msTick();
+        processChildren(scriptContext, targetContext, sourceContext);
+        unsigned end = msTick();
+        unsigned delta = (end >= start ? end - start : UINT_MAX - start + end);
+        switch (m_mode)
+        {
+        case Append:
+            if (!txSummary->append(m_info.m_name, delta, m_info.m_level, m_info.m_groups))
+                UWARNLOG("script operation %s did not append '%s=%ums'; name is malformed or is in use", m_tagname.str(), m_info.m_name.str(), delta);
+            break;
+        case Set:
+            if (!txSummary->set(m_info.m_name, delta, m_info.m_level, m_info.m_groups))
+                UWARNLOG("script operation %s did not set '%s=%ums'; name is malformed or identifies a cumulative timer", m_tagname.str(), m_info.m_name.str(), delta);
+            break;
+        case Accumulate:
+            if (!txSummary->updateTimer(m_info.m_name, delta, m_info.m_level, m_info.m_groups))
+                UWARNLOG("script operation %s did not add %ums to cumulative timer '%s'; name is malformed or identifies a scalar value", m_tagname.str(), delta,  m_info.m_name.str());
+            break;
+        default:
+            break;
+        }
+        return true;
+    }
+
+    virtual void toDBGLog () override
+    {
+    #if defined(_DEBUG)
+        DBGLOG(">>>%s> %s %s >>>>", m_traceName.str(), m_tagname.str(), m_info.m_name.str());
+        CEsdlTransformOperationWithChildren::toDBGLog();
+        DBGLOG (">>>>>>>>>>> %s >>>>>>>>>>", m_tagname.str());
+    #endif
+    }
+};
+
+class CEsdlTransformOperationDelay : public CEsdlTransformOperationWithoutChildren
+{
+protected:
+    unsigned m_millis = 1;
+public:
+    CEsdlTransformOperationDelay(IXmlPullParser& xpp, StartTag& stag, const StringBuffer& prefix)
+        : CEsdlTransformOperationWithoutChildren(xpp, stag, prefix)
+    {
+        const char* millis = stag.getValue("millis");
+        if (!isEmptyString(millis) && s_deserializer.deserialize(millis, m_millis) != Deserialization_SUCCESS)
+            esdlOperationError(ESDL_SCRIPT_InvalidOperationAttr, m_tagname, "invalid millis", m_traceName, !m_ignoreCodingErrors);
+    }
+
+    virtual ~CEsdlTransformOperationDelay()
+    {
+    }
+
+    virtual bool exec(CriticalSection* crit, IInterface* preparedForAsync, IEsdlScriptContext* scriptContext, IXpathContext* targetContext, IXpathContext* sourceContext) override
+    {
+        OptionalCriticalBlock block(crit);
+        unsigned then = msTick();
+        while (msTick() - then < m_millis) {}
```

Comment:
This looks pretty expensive.  Why not just use the jlib function "MilliSleep"?  You can use your test cases to see if it solves your issue of occasional misses.  You could also compare CPU usage.

Reply:
I wasn't aware of MilliSleep and, considering that the unit tests may be the only place the operation is ever used, I wasn't too concerned about finding it. It does not solve the occasional miss; it makes it more likely  (based on my observation of a 100% miss rate, up from 0%, on my local machine).

It is, however, friendlier to systems where the churn could translate into financial cost. If the misses can't be prevented, I higher miss rate may be preferable to a less CPU intensive implementation.